File: documentation/guides/rust/ockam_get_started/src/hop.rs
use ockam::{Any, Context, Result, Routed, Worker};

pub struct Hop;

#[ockam::worker]
impl Worker for Hop {
    type Message = Any;
    type Context = Context;

    /// This handle function takes any incoming message and forwards
    /// it to the next hop in its onward route
    async fn handle_message(&mut self, ctx: &mut Context, msg: Routed<Any>) -> Result<()> {
        println!("Address: {}, Received: {}", ctx.address(), msg);

        let mut msg = msg.into_transport_message();
        msg.onward_route.step()?;
        msg.return_route.modify().prepend(ctx.address());
        ctx.forward(msg).await
    }
}

File: implementations/elixir/ockam/ockam/lib/ockam/examples/hop.ex
defmodule Ockam.Examples.Hop do
  @moduledoc false

  use Ockam.Worker

  alias Ockam.Message
  alias Ockam.Worker

  @impl true
  def handle_message(message, state) do
    Worker.route(Message.forward(message) |> Message.trace(state.address), state)

    {:ok, state}
  end
end
